### ISSUE TITLE
Add API to fix violations automatically

### ISSUE THREAD

**Issue by vlsi** [checkstyle/checkstyle#7427]
https://checkstyle.sourceforge.io/config_whitespace.html#WhitespaceAround

Sample output:

```
[ant:checkstyle] [ERROR] /../calcite/core/src/main/java/org/apache/calcite/plan/RelOptUtil.java:3527:53: ':' is not preceded with whitespace. [WhitespaceAround]
[ant:checkstyle] [ERROR] /../calcite/core/src/main/java/org/apache/calcite/plan/RelOptUtil.java:3528:54: ':' is not preceded with whitespace. [WhitespaceAround]
```

Just in case, the code in question was

```java
            joinConvention == Convention.NONE ? left: RelOptRule.convert(left, joinConvention),
            joinConvention == Convention.NONE ? right: RelOptRule.convert(right, joinConvention),
```

Even though the message is more or less clear, the case is trivial, and Checkstyle could insert the space automatically.

The same goes for UnusedImports, ImportOrder, CustomImportOrder, RedundantImport, ModifierOrder, and so on.

Adding auto-correction API to Checkstyle would save a lot of time for developers across the world.

Of course, there always will be checks that can't be automatically corrected. For instance, AnonInnerLength requires manual input for resolution.

Note: the purpose of this issue is not not make all the checks auto-correctable, but I suggest an API is added, so more and more checks could become auto-correctable in the future.

Note: [ktlint](https://github.com/pinterest/ktlint) implements auto-correctable checks for Kotlin, and it is really nice to use.

**Comment by strkkk** [comment]
It is good idea, but sounds more like a feature for IDE plugins.

**Comment by romani** [comment]
This is big project, and it is almost blocked by incomplete xpath support in Checks.
As all Checks become xpath compliant, auto correction become good point to discuss.

It is good to look around of tools who already do autocorrections (IDEs, tools,  and do something similar). Visual studio checkstyle plugin already have some auto correction so also need to be reviewed.

**Comment by vlsi** [comment]
Frankly speaking, I do not understand why auto-correct feature is "blocked by incomplete xpath support in Checks". Can you please clarify?

I fully agree that UI (and/or command-line tools) for auto-corrections are out of scope for Checkstyle, however, Checkstyle itself should provide an API so the tools (e.g. IDEs or build system plugins) could run Checkstyle and get the set of correctable errors, non-correctable errors and the suggested corrections.

Let us take "import order" for instance.
`ktlint` is able to arrange Kotlin imports automatically, and they can return the source with the arranged import order.
Then that Java API can be used in build system tools (for instance, https://github.com/autostyle/autostyle/ ), so the end-users can see the suggested changes and issue a command to apply them.

In the same way, https://github.com/google/google-java-format provides an automatic import order format for Java. In the same way, Autostyle plugin for Gradle uses google-java-format so the end-users can get auto-correctable messages.

Here's an example from the current master branch of Apache Calcite. I intentionally added some violations (extra line break, unused import, and shuffled two imports a bit).
The error message is very understandable, and it prints a command to correct the violations.

```
./gradlew autostyleCheck

> The following files have format violations:
    core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableMergeJoin.java
      @@ -21,17 +21,15 @@
       import org.apache.calcite.linq4j.tree.Expression;
       import org.apache.calcite.linq4j.tree.Expressions;
       import org.apache.calcite.linq4j.tree.ParameterExpression;
      -import java.util.Random;
      -
       import org.apache.calcite.plan.RelOptCluster;
       import org.apache.calcite.plan.RelOptCost;
      -import org.apache.calcite.rel.RelNodes;
       import org.apache.calcite.plan.RelOptPlanner;
       import org.apache.calcite.plan.RelTraitSet;
      -import org.apache.calcite.rel.RelCollationTraitDef;
       import org.apache.calcite.rel.RelCollation;
      +import org.apache.calcite.rel.RelCollationTraitDef;
       import org.apache.calcite.rel.RelCollations;
       import org.apache.calcite.rel.RelNode;
      +import org.apache.calcite.rel.RelNodes;
       import org.apache.calcite.rel.core.CorrelationId;
       import org.apache.calcite.rel.core.Join;
       import org.apache.calcite.rel.core.JoinRelType;
  Run './gradlew autostyleApply' to fix the violations.
```

Just in case: Autostyle implements Gradle plugin with autostyleCheck and autostyleApply commands, it renders the diff for violations. And it uses Google Java Format library to arrange imports.

I want something like that from Checkstyle as well. WDYT?

**Comment by romani** [comment]
No special API is required to make automated correction, that is why it is already exists in some plugins. It might be not very convenient, and this is what we will investigate in this issue.

>I do not understand why auto-correct feature is "blocked by incomplete xpath support in Checks". Can you please clarify?

Some Checks report violation on line only (not a column and line) so human detect issue easily, but auto correction will struggle a bit.

>I want something like that from Checkstyle as well. WDYT?

Some custom Check should be created that takes report with violations and traverse tree again and try to fix. I just try to avoid any public API for this now, as we do not have experience in this ( we do not know all nuances and problems). I do not want to release contract (API) to plugins now. Somebody needs to investigate this and provide report on this.
In other words let's have some implementation first( or review plugins or ...), test it on real life code, and then generalize to API.

**Comment by Fernal73** [comment]
As far as I can see (without getting into technical details such as XPath), it's a question of investigating Checkstyle rules as to which can be auto-corrected within the existing limitations of the Checkstyle API.

Where it becomes tricky is with the myriad options that a rule has and thus myriad solutions. This, too, is doable if there's a one-to-one mapping from a violation to the desired correction. When there's more than one solution , the solution needs to be specified explicitly somewhere. Or only one best solution provided.

E.g.,
Static classes can be defined as utility class by labeling the class final and providing a private constructor that throws an exception when invoked. Or define the class as an empty enum. 

Wrt to formatting violations, it's best to run the desired formatting rules prior to Checkstyle to minimise violations and then try for auto-correcting via the Checkstyle API. 
My prognosis that there will be fewer auto-correcting of the formatting variety needed this way.

**Comment by Fernal73** [comment]
Additionally, wouldn't the auto-correct feature have to provide the user the option of suppressing the violation?

**Comment by vlsi** [comment]
@Fernal73 , I'm afraid you are overcomplicating things.

There's exactly one way to fix `UnusedImports` violation.
Checkstyle had `UnusedImports` for ages, `UnusedImports` check is trivial for auto-correction.

For now, it looks like the best approach is to avoid Checkstyle when possible (e.g. imports, modifier orders, etc, etc).

**Comment by Fernal73** [comment]
My mistake. The title states that it's about applying corrections automatically.
I wasn't aware that you were intent on limiting the scope to the issues you highlighted.

I was thinking more on the lines of something like this:

https://blog.atomist.com/when-robots-can-do-all-the-tedious-work/

**Comment by Fernal73** [comment]
If this post is about implementing trivial checks only, how about a Checkstyle rule that allows users to specify if single line while, if or for loops are to be split across lines or not. It's not supported by Google Java format or Prettier but it's something some developers like me would prefer. Curly braces, be damned. That would be a trivial auto-correction as well---if there was a rule for it, that is.

**Comment by vlsi** [comment]
Here are sample configuration files: https://github.com/apache/calcite/blob/master/src/main/config/checkstyle/checker.xml
 , https://github.com/apache/jmeter/blob/master/config/checkstyle/checkstyle.xml

Both projects used Checkstyle for ages, and now it looks like the way to go is to migrate as many rules as it is possible to google-java-format or something like that.

> Checkstyle rule that allows users to specify if single line while, if or for loops are to be split across lines or not

I don't think I have seen such code styles, so I am not comfortable discussing that.

**Comment by Fernal73** [comment]
My mistake. I ought to have provided you the context of my remark:

https://github.com/jhipster/prettier-java/issues/337

https://github.com/prettier/prettier/issues/7141

Your fix appears to be about running API calls to Google Java format after Checkstyle flags the error. 
I've encountered similar problems. I've found the simplest solution is to run the formatter first and then run Checkstyle against the formatted input. I simply let the formatter overwrite the existing code in-place. You seem to be suggesting a manual selection of changes to be accepted or rejected.

**Comment by vlsi** [comment]
>Your fix appears to be about running API calls to Google Java format after Checkstyle flags the error.

No.

There's [Autostyle Gradle plugin](https://github.com/autostyle/autostyle/) that integrates multiple formatters/linters. It would be great if Checkstyle could be integrated there as well.

> Are you suggesting that Checkstyle

No. I suggest that certain Checksyle checks should come with integrated auto-correction available.
For instance: "imprt order check". Checkstyle has import order check, and it should come with ability to auto-correct wrong imports by default.
Note: "import order" is configurable in Checkstyle, so if you suggest running other formatter first, then I would have to configure the same import orders in two places.

I don't like to configure import order in two places, so I removed "ImportOrder / UnusedImport" from Checkstyle configuration and replaced it with Autostyle's import order configuration (it delegates format/check to google-java-format).

**Comment by Fernal73** [comment]
I'm not familiar with Gradle since I use Ant instead and build scripts.
Actually, ESLint is the linter and Prettier, the formatter. Prettier-ESLint fixes prettier output as per corrections configured. That's what the documentation states, anyway. It didn't work for Java source but I expect it does for JavaScript. 

I'm fine with auto-correction available for Checkstyle rules. Nothing wrong with that.

**Comment by Fernal73** [comment]
The other way to approach this as @strkkk suggested is to leave it to external tools to fix such as Prettier does via the Prettier-ESlint plugin. Checkstyle can be used to provide the rules for auto-correction. Tools might prefer a more lightweight file format such as json or yaml. 
There's another discussion wrt that:

https://github.com/checkstyle/checkstyle/issues/7246

**Comment by romani** [comment]
there are some autofixes implementation for Checks violations at http://unkrig.de/w/Cs-contrib.unkrig.de , project owned by @aunkrig .

**Comment by vlsi** [comment]
:)

This is very relevant: https://elm-lang.org/news/compiler-errors-for-humans

**Comment by Fernal73** [comment]
You can try this as well:

https://spacevim.org/

**Comment by vlsi** [comment]
For what?

**Comment by Fernal73** [comment]
https://spacevim.org/use-vim-as-a-java-ide/

Linting and formatting.

**Comment by vlsi** [comment]
Thank you, I'm fine with IntelliJ IDEA.

**Comment by Fernal73** [comment]
https://spacevim.org/layers/lang/elm/
Also, integrated with Elm.
Except that the npm packages are not installable on Android.

**Comment by romani** [comment]
Investigate one more tool with autofix https://twitter.com/jon_k_schneider/status/1230500526250496001?s=19

**Comment by landsman** [comment]
Yes please, look on Javascript world - Prettier tool. Easy, comfortable connected to Husky. Developer is super happy, without thinking about checkstyle.

**Comment by Hrushi20** [comment]
Hey! I would love to work on this project. Can anybody help me out in this project. I am new to open source development

**Comment by Tilakraj0308** [comment]
Can someone help me out with this issue cause I am new to open source development and want a headstart.

**Comment by Tilakraj0308** [comment]
@romani can you plz help me out to understand the issue cause I am new to open source development

**Comment by romani** [comment]
@Tilakraj0308, please start from some another issue, this issue is huge project.

Looks at issues labelled by "first gsoc issue".

**Comment by Tilakraj0308** [comment]
> @romani , thankyou very much

**Comment by Tiku24** [comment]
Hey I am a new beginner can you suggest me how I can start at initial

**Comment by rnveach** [comment]
@Tiku24 I do not suggest this issue for someone new. Please read https://github.com/checkstyle/checkstyle/wiki/To-student-of-GSoC-2020 .

**Comment by developernarendra** [comment]
hi i am new here how can i start this project pls help someone

**Comment by rnveach** [comment]
@developernarendra Since you are new, find another issue with an easy label.

**Comment by rnveach** [comment]
A possible example of what auto-fixing may need to entail: https://github.com/rnveach/checkstyle/commits/truth_replacer

We will run into issues with order of updates, conflicting updates, and updates that need their own updates.

**Comment by gquemener** [comment]
I have discovered that openrewrite provides a degree of integration with checkstyle.
AFAIU, openrewrite will read the checkstyle.xml to configure its own recipes.

However, base violations (for example, missing whitespace and wrong import orders) are not automatically fixed in my experience.

A lead, in order to provide a violation fix feature, could be to investigate a collaboration with this other tool, couldn't it?

**Comment by vlsi** [comment]
I incline it is better to use https://editorconfig.org/ as a common ground for the configuration.
For instance, it has `insert_final_newline`, and it is understood by editors and linters, so it would probably make sense to make checkstyle read and understand editorconfig.

**Comment by johnameyer** [comment]
I see this is 3 years old but still "ongoing discussion" - is there any consensus on this? As others have mentioned autoformat capabilities are a basic functionality of linters these days and pushing the responsibility onto editor plugins leaves terminal editors excluded

**Comment by romani** [comment]
No disagreement on need,we just need find people to design and to do this.

**Comment by Zaljumaei** [comment]
@romani ; I want to contribute with you to this problem for GSOC23.  I will try to understand the problem to make the first contribution and have a chance to get selected in GSOC23

**Comment by romani** [comment]
@Zaljumaei, please focus on resolving simple issues first as described at https://github.com/checkstyle/checkstyle/wiki/To-GSoC-participant

**Comment by rnveach** [comment]
https://github.com/checkstyle/checkstyle/compare/master...rnveach:checkstyle:truth_replacer is an example of may what be needed to auto fix violations for checks. This example doesn't represent the full API changes needed to CS, just the basic and flat idea.

Violations have to be processed in reverse order on the FileText file and saved to the file. We will need to know all violations in this case as multiple checks can put violations on the same file, and modifying one violation will change the location of other violations. Reverse order ensured I didn't have to recalculate other violations as I assume file will be made larger after the violation. Things may get more complicated if there is multiple violations on the same line and definitely same column position.

All necessary information of the violation may need to be saved in order to know the context of the violation and apply the fix. This will be dependent on each check and may require additional information not found in the violation message. In my example, I needed to know the method being violated on and the parameters of the method to know how to do the transformation.

**Comment by Deltachaos** [comment]
Coming from PHP its real fascinating that there is even a discussion if this is a feature you want or not. [PHP_CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer) does this for years already and saved hundreds of thousands of developer hours. I found myself few days ago reordering import statements for two hours which could easily be automated.

**Comment by romani** [comment]
no api extension is planned, 
we created project https://github.com/checkstyle/checkstyle-openrewrite-recipes 
to allow auto fixes based on regular violation report.
